sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -77,20 +77,28 @@ public ShardRecordsIterator(final ShardCheckpoint initialCheckpoint,
 
   private void readMoreIfNecessary() throws TransientKinesisException {
     if (data.isEmpty()) {
-      GetKinesisRecordsResult response;
+      GetKinesisRecordsResult response = fetchRecords();
+      data.addAll(filter.apply(response.getRecords(), checkpoint));
+    }
+  }
+
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    GetKinesisRecordsResult response = null;
+    do {
       try {
         response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
             checkpoint.getShardId());
+        shardIterator = response.getNextShardIterator();
       } catch (ExpiredIteratorException e) {
         LOG.info("Refreshing expired iterator", e);
         shardIterator = checkpoint.getShardIterator(kinesis);
-        response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
-            checkpoint.getShardId());
       }
-      LOG.debug("Fetched {} new records", response.getRecords().size());
-      shardIterator = response.getNextShardIterator();
-      data.addAll(filter.apply(response.getRecords(), checkpoint));
-    }
+    } while (response == null || isEmptyResponseBeforeStreamEnd(response));
+    return response;
+  }
+
+  private boolean isEmptyResponseBeforeStreamEnd(GetKinesisRecordsResult response) {
```

Comment:
this name doesn't tell me much. Maybe something like:
notEnoughRecordsPulled?

Reply:
Current name also emphasises that we still haven't catch up with the stream but somehow we got the empty response (it can really happen). I would rather stay with current name. What do you think?

Reply:
So maybe simply: gotEmptyResponseButIsBeforeEndOfTheStream

Reply:
Ok, I like that